serverless-architecture-on-aws/lab-3/README.md

```diff
@@ -0,0 +1,459 @@
+    role: role.arn,
+    code: new pulumi.asset.AssetArchive({
+        ".": new pulumi.asset.FileArchive("./app")
+    }),
+    handler: "s3.handler",
+    runtime: aws.lambda.Runtime.NodeJS16dX
+});
+
+const logGroup = new aws.cloudwatch.LogGroup("loggroup", {
+    name: pulumi.interpolate`/aws/lambda/${fn.name}`
+});
+
+const bucket = new aws.s3.Bucket("lambdaBucket");
+bucket.onObjectCreated("onObjectCreated", fn);
+
+export const bucketName = bucket.bucket;
+```
+</details>
+
+### CloudWatch Schedule
+
+If you've not used this before it's very useful because you can effectively create CRON jobs that run your Lambda Functions.
+
+Same as before, here's the Lambda Function code:
+
+```javascript
+exports.handler =  async function(event, context) {
+    const cloudwatchEvent = event;
+    const logJSON = {
+        id: event.id,
+        timestamp: event.time,
+        eventArn: event.resources[0]
+    };
+    console.log(logJSON)
+}
+```
+
+Here's the Lambda Function resource with the handler input changed:
+
+```typescript
+const fn = new aws.lambda.Function("function", {
+    role: role.arn,
+    code: new pulumi.asset.AssetArchive({
+        ".": new pulumi.asset.FileArchive("./app")
+    }),
+    handler: "cloudwatch.handler", // changed from "s3.handler"
+    runtime: aws.lambda.Runtime.NodeJS16dX
+});
+```
+
+The code for the CloudWatch Scheduler resource:
+
+```typescript
+const scheduler = new aws.cloudwatch.EventRule("scheduler", {
+    scheduleExpression: "rate(1 minute)"
+});
+```
+
+and to wire it up to the Lambda Function:
+
+```typescript
+scheduler.onEvent("schedulerevent", fn)
+```
+
+As we're not going to use the CLI to run the scheduler, we don't need to export any outputs.
+
+Then an update:
+
+```zsh
+pulumi up
+```
+
+<details><summary>The complete CloudWatch Scheduler Pulumi code can be seen here</summary>
+
+```typescript
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+
+const role = new aws.iam.Role("role", {
+    assumeRolePolicy: aws.iam.assumeRolePolicyForPrincipal(aws.iam.Principals.LambdaPrincipal),
+    managedPolicyArns: [
+        aws.iam.ManagedPolicy.AWSLambdaBasicExecutionRole, 
+    ]
+});
+
+const fn = new aws.lambda.Function("function", {
+    role: role.arn,
+    code: new pulumi.asset.AssetArchive({
+        ".": new pulumi.asset.FileArchive("./app")
+    }),
+    handler: "cloudwatch.handler",
+    runtime: aws.lambda.Runtime.NodeJS16dX
+});
+
+const logGroup = new aws.cloudwatch.LogGroup("loggroup", {
+    name: pulumi.interpolate`/aws/lambda/${fn.name}`
+});
+
+const scheduler = new aws.cloudwatch.EventRule("scheduler", {
+    scheduleExpression: "rate(1 minute)"
+});
+
+scheduler.onEvent("schedulerevent", fn)
+```
+</details>
+
+## Tidying up
+
+I think pretty much everything here is covered by the AWS Free Tier, but if you want to keep things neat and tidy, then all that is left is to run:
+
+```zsh
+pulumi destroy
+``` 
+
+to delete everything.
```

Comment:
Below this step, I would link back to the serverless templates.  If you are doing it somewhere else, that is fine. Something like this:

## SHORTCUT - Pulumi Templates
[Pulumi Templates](https://www.pulumi.com/templates/?utm_source=da&utm_medium=referral&utm_campaign=workshops&utm_content=ced-fall2022-workshops) are the fastest way to deploy infrastructure to AWS, Azure, and Google Cloud. After deploying, you can easily modify the infrastructure by updating the code in TypeScript, Python, Go, Java, .NET/C#, or YAML.

This entire workshop can be found at [AWS Serverless Application](https://www.pulumi.com/templates/serverless-application/aws/?utm_source=da&utm_medium=referral&utm_campaign=workshops&utm_content=ced-fall2022-workshops) template to:

```bash
mkdir my-serverless-app && cd my-serverless-app
pulumi new serverless-aws-typescript -y
pulumi up -y
open $(pulumi stack output siteURL)
```

Remember to clean up your resources
```bash
pulumi destroy -y
```

Reply:
I think they'll be fine finding their own way back to the start if they need to